MonoDevelop.MSBuild.Editor/Completion/MSBuildCompletionSource.cs

```diff
@@ -119,19 +125,141 @@ CompletionContext CreateCompletionContext (List<CompletionItem> items)
+				result = await packageResponse.Content.ReadAsStringAsync ();
+
+			} catch (Exception) {
+				return CompletionContext.Empty;
+			}
+
+			var packageJSONObject = JObject.Parse (result);
+			var packageArray = (JArray)packageJSONObject["data"];
+
+			foreach (var package in packageArray) {
+				var info = package.ToString ();
+				items.Add (CreateCompletionItem (new ItemInfo (info, null, null),
+						MSBuildSpecialCommitKind.AttributeValueSpecial));
+			}
+			return new CompletionContext (ImmutableArray<CompletionItem>.Empty.AddRange (items));
+		}
+
+		async Task<CompletionContext> GetPackageVersionsAsync (MSBuildRootDocument doc,
+																MSBuildResolveResult rr)
+		{
+			if (rr == null) {
+				return CompletionContext.Empty;
+			}
+
+			var items = new List<CompletionItem> ();
+			var httpClient = new HttpClient (); ;
+			string result = null;
+			var packageName = rr.XElement.Attributes.First.Value.ToLower ();
+
+			if (string.IsNullOrEmpty (packageName)) {
+				return CompletionContext.Empty;
+			}
+
+			try {
+				//includes pre-release: https://docs.microsoft.com/en-us/nuget/api/search-query-service-resource
+				var cancellationVersion = new CancellationTokenSource (TimeSpan.FromSeconds (100));
+				var versionResponse = await httpClient.GetAsync ($"https://api-v2v3search-0.nuget.org/query?q={packageName}&prerelease=true", cancellationVersion.Token);
+				result = await versionResponse.Content.ReadAsStringAsync ();
+
+			} catch (Exception) {
+				return CompletionContext.Empty;
+			}
+
+			var versionJSONObject = JObject.Parse (result);
+			var versionArray = (JArray)versionJSONObject["data"];
+
+			//var totalHits = (int)versionJSONObject["totalHits"];
+
+			//Normally, it would be versionArray[0]->id == packageName
+			//If it doesn't, the for loop will make sure that correct versions of that exact packageName are listed
+			foreach (var version in versionArray) {
+				var id = version["id"].ToString ().ToLower ();
+				if (id == packageName.ToLower ()) {
+					var description = version["description"].ToString ();
+					foreach (var ver in version["versions"]) {
+						items.Add (CreateCompletionItem (new ItemInfo (ver["version"].ToString (), description, description),
+							MSBuildSpecialCommitKind.AttributeValueSpecial));
+					}
+					break;
+				}
+			}
+
+			return new CompletionContext (ImmutableArray<CompletionItem>.Empty.AddRange (items));
+		}
+
+		CompletionItem CreateCompletionItem (BaseInfo info, MSBuildSpecialCommitKind kind, string prefix = null)
 		{
 			var image = DisplayElementFactory.GetImageElement (info);
 			var item = new CompletionItem (prefix == null ? info.Name : prefix + info.Name, this, image);
 			item.AddDocumentationProvider (this);
-			item.Properties.AddProperty (typeof(BaseInfo), info);
+			item.Properties.AddProperty (typeof (BaseInfo), info);
+			item.Properties.AddProperty (this, kind);
```

Comment:
I would key this on `typeof(MSBuildSpecialCommitKind)`

Reply:
Resolved in my latest [commit ](https://github.com/mhutch/MonoDevelop.MSBuildEditor/pull/19/commits/c3310c63800989d3715b58f93ea6db91bc73db12). Thanks!